### PR TITLE
Various simplifications, optimizations and features related to clipping.

### PR DESCRIPTION
**Pull request by glennw**
* Introduce ClipStore, an arena for storing clip sources.
    This reduces a lot of memory copying and allocations, since the clip stack built each frame now just contains handles into the clip store, rather than copying the entire mask cache.
* Move remaining mask cache code into ClipSources, and simplify it.
    We no longer need to copy the border sources etc into the mask cache structures - they can be referenced directly from the clip sources.
* Use a separate GPU cache handle for each clip, so ClipAddressRange code is no longer needed.
    This is possible since we use the GpuCache now, which doesn't need to allocate space during initialization.
* Simplify the code in the clip batcher that adds clip instances to the render targets.
* Add support for multiple image masks (a by-product of removing the init step of the mask cache code).
* Move MaskBounds into the ClipSources (from the MaskCacheInfo struct).
* Add recycle() to free list, and support cloning weak handles.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/webrender/1679)

### CODE REVIEW

**Comment by glennw** [comment]
r? @kvark 

https://hg.mozilla.org/try/rev/c67da9f751701d22a7a7b7c2b9ea4d024b99f661
https://treeherder.mozilla.org/#/jobs?repo=try&revision=c67da9f751701d22a7a7b7c2b9ea4d024b99f661

**Comment by glennw** [comment]
This change is probably easier to review by considering each commit as a separate improvement...

File: webrender/src/clip.rs [hunk 1]
```diff
@@ -109,18 +109,16 @@ impl From<ClipRegion> for ClipSources {
 
 #[derive(Debug)]
 pub struct ClipSources {
-    clips: Vec<ClipSource>,
-    pub mask_cache_info: MaskCacheInfo,
+    pub clips: Vec<ClipSource>,
+    pub handles: Vec<GpuCacheHandle>,
```
Comment (kvark) [inline webrender/src/clip.rs]: why don't we want to either simply include `GpuCacheHandle` into `ClipSource` or have it stored nearby, e.g. `Vec<(ClipSrouce, GpuCacheHandle)>`?

Comment (glennw) [reply webrender/src/clip.rs]: I did consider including it in ```ClipSource``` but it seemed a bit cleaner conceptually to separate them. No real reason it couldn't be in a tuple though!

Comment (kvark) [reply webrender/src/clip.rs]: Great, let's tuple it up then ;)

**Comment by glennw** [comment]
@kvark OK, pushed a commit that moves ```handles``` into a tuple in ```clips```.

**Comment by kvark** [comment]
Thank you!
@bors-servo r+

**Pull request merged**